### ISSUE TITLE
gem installation failure with clang and gcc

### ISSUE THREAD

**Issue by hsbt** [rubygems/rubygems#8015]
### Describe the problem as clearly as you can

When users use pre-compiled ruby provided by rvm, some issues happend with compiler flag differences.

* https://ml.ruby-lang.org/archives/list/ruby-talk@ml.ruby-lang.org/message/OI3BLGJEEZHEFHAYMCNU7AM5BV465R6T/
* https://github.com/rubygems/rubygems/issues/1508#issuecomment-2336873801
* https://github.com/ruby/bigdecimal/issues/297#issuecomment-2328928055

```
gcc: error: unrecognized command line option ‘-fdeclspec’
gcc: error: unrecognized command line option ‘-Wextra-tokens’; did you mean ‘-Wextra-semi’?
gcc: error: unrecognized command line option ‘-Wdivision-by-zero’; did you mean ‘-Wdiv-by-zero’?
gcc: error: unrecognized command line option ‘-Wshorten-64-to-32’
```

`extconf.rb` or RubyGems build toolchain use `RbConfig::CONFIG` by pre-compiled binary built by `clang`. But it's not working wigh `gcc`.

I'm not sure what we resolve this issue yet.

**Comment by deivid-rodriguez** [comment]
People at https://github.com/ruby/bigdecimal/issues/297 report that this issue was introduced with Ruby 3.3.4. Is that correct? Do we know which specific change?

**Comment by hsbt** [comment]
In my understanding, this report is not related the recent releases of Ruby 3.3.

@nobu said `RbConfig::CONFIG['CC']` don't use missing compiler like `/usr/bin/clang`, but their report try to use `gcc` for that. I'm not sure what happen their environment.

**Comment by askrynnikov** [comment]
Hi!
Development machine, Ubuntu 20.04.6 LTS, using rvm, the problem occurred when upgrading from ruby ​​3.3.3 to 3.3.4 (3.3.5)

1. reinstalling rvm did not help
2. rebooting the computer did not help
3. installing libjemalloc-dev libgmp-dev build-essentials did not help
4. upgrading gcc from version 9.4.0 to version 11.4.0 did not help

`sudo nano /home/andrey/.rvm/gems/ruby-3.3.4/extensions/x86_64-linux/3.3.0/bigdecimal-3.1.8/mkmf.log`
Contents of the mkmf.log file:

> LD_LIBRARY_PATH=.:/home/andrey/.rvm/rubies/ruby-3.3.4/lib "/usr/bin/gcc -o conftest -I/home/andrey/.rvm/rubies/ruby-3.3.4/include/ruby-3.3.0/x86_64-linux -I/home/andrey/.rvm/rubies/ruby-3.3.4/include/ruby-3.3.0/ruby/backward -I/home/andrey/.rvm/rubies/ruby-3.3.4/include/ruby-3.3.0 -I. -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-d>
> gcc: error: unrecognized command-line option ‘-fdeclspec’
> gcc: error: unrecognized command-line option ‘-Wextra-tokens’; did you mean ‘-Wextra-semi’?
> gcc: error: unrecognized command-line option ‘-Wdivision-by-zero’; did you mean ‘-Wdiv-by-zero’?
> gcc: error: unrecognized command-line option ‘-Wshorten-64-to-32’
> checked program was:
> /* begin */
> 1: #include "ruby.h"
> 2:
> 3: int main(int argc, char *argv)
> 4: {
> 5: return !!argv[argc];
> 6: }
> / end */

Only the solution from @martinnicolas helped

```
sudo apt install clang
export CC=/usr/bin/clang
export CXX=/usr/bin/clang++
```

P.S. Used in production rbenv, haven't tried updating 3.3.3 -> 3.3.5 yet

**Comment by deivid-rodriguez** [comment]
Do we have a way to reproduce this?

**Comment by martinnicolas** [comment]
@deivid-rodriguez I think that you will be able to reproduce this by following this steps:

1 - Install rvm 1.29.12

2 - Install ruby 3.3.4

3 - Install [bigdecimal](https://github.com/ruby/bigdecimal) gem

	gem install bigdecimal

**Comment by hsbt** [comment]
And Linux with `--binary` installation.

**Comment by nobu** [comment]
This is because RVM replaces the `CC` value, if that command is not found at installation time.

https://github.com/rvm/rvm/blob/master/scripts/functions/build#L68-L87

While RVM installs `gcc` as a required package and falls back to `"/usr/bin/gcc"`, the binaries are built with `clang`.

By installing `clang` before `rvm install --binary`, `gem install` worked successfully even without the `CC` environment variable.

**Comment by hsbt** [comment]
Thanks. There is no action for rubygems and ruby sides.

@martinnicolas I recommend to not use rvm and their binary installation.

**Comment by deivid-rodriguez** [comment]
Nice investigation @nobu. Should RVM stop doing this, maybe?

**Comment by nobu** [comment]
@deivid-rodriguez I agree with that, since not only `CFLAGS`, also config.h need to match `CC`.

Or, it should install the compiler same as pre-built binaries, at least.

**Comment by joao-esteves** [comment]
> Only the solution from @martinnicolas helped
> 
> ```
> sudo apt install clang
> export CC=/usr/bin/clang
> export CXX=/usr/bin/clang++
> ```
> 
> P.S. Used in production rbenv, haven't tried updating 3.3.3 -> 3.3.5 yet

This didn't work for me in Ubuntu 24.04.1 LTS with RVM, Ruby 3.3.5, and the latest Clang through APT. It fixed the reported errors but then I had more, e.g. not finding <cstdbool> (due to being removed in C++20?).

What worked is fully removing RVM (not just the Rubies) through `rvm implode`, clearing some things it couldn't delete with `sudo rm -rf ~/.rvm`, and reinstalling RVM. I no longer had Clang during all of this, it was able to use the built-in GCC 13.2.

**Comment by martinnicolas** [comment]
@joao-esteves Thanks! I'll consider that.. happy programmer days! 😃

**Comment by istrasci** [comment]
> What worked is fully removing RVM (not just the Rubies) through `rvm implode`, clearing some things it couldn't delete with `sudo rm -rf ~/.rvm`, and reinstalling RVM. I no longer had Clang during all of this, it was able to use the built-in GCC 13.2.

This worked for me as well!

**Comment by dangquangthai** [comment]
> And Linux with `--binary` installation.

```shell
rvm uninstall 3.3.4
rm -rf ~/.rvm/gems/ruby-3.3.4*
rm -rf ~/.rvm/rubies/ruby-3.3.4/

rvm install 3.3.4 --binary
bundle install
```

worked perfectly!

**Comment by gazeldx** [comment]
In Amazon Linux, I solved it by `yum install ruby-devel`.

**Comment by lfalcao** [comment]
it works with `rvm install 3.0.7 --disable-binary`.
